### PR TITLE
Optimize get_objects_inside_radius calls

### PR DESCRIPTION
**Pull request by nerzhul**
our previous implementation calls the ActiveObjectMgr to return ids and then lookup those ids in the same map and test each object
Instead now we call the global map to return the pointers directly and we ask filtering when building the list using lamba.

This drop double looping over ranges of active objects (and then filtered one) and drop x lookups on the map regarding the first call results

## To do

This PR is a Ready for Review.

## How to test

Have item_drops and a mob framework with mobs and just see it works (a little bit better)

### CODE REVIEW

File: src/collision.cpp [hunk 1]
```diff
@@ -360,17 +360,19 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
...
 
-				for (u16 obj_id : s_objects) {
-					ServerActiveObject *current = s_env->getActiveObject(obj_id);
-
-					if (!self || (self != current &&
-							self != current->getParent())) {
-						objects.push_back((ActiveObject*)current);
+				// search for objects which are not us, or we are not its parent
+				// we directly use the callback to populate the result to prevent
+				// a useless result loop here
```
Comment (sfan5) [inline src/collision.cpp:366]: Wouldn't passing `objects` to `getObjectsInsideRadius` work?

File: src/collision.cpp [hunk 2]
```diff
@@ -360,17 +360,19 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
...
+				auto includeObjCallback = [self, &objects](ServerActiveObject *obj) {
```
Comment (sfan5) [inline src/collision.cpp:367]: variables are generally snake_case, also spacing between [] and ()

File: src/server/activeobjectmgr.cpp [hunk 3]
```diff
@@ -111,17 +111,19 @@ void ActiveObjectMgr::removeObject(u16 id)
 }
 
 // clang-format on
-void ActiveObjectMgr::getObjectsInsideRadius(
-		const v3f &pos, float radius, std::vector<u16> &result)
+void ActiveObjectMgr::getObjectsInsideRadius(const v3f &pos, float radius,
+		std::vector<ServerActiveObject *> &result,
+		std::function<bool(ServerActiveObject *obj)> includeObjCallback)
```
Comment (sfan5) [inline src/server/activeobjectmgr.cpp:116]: snake_case here too

File: src/serverenvironment.cpp [hunk 4]
```diff
@@ -1603,14 +1603,12 @@ void ServerEnvironment::getSelectedActiveObjects(
 	const core::line3d<f32> &shootline_on_map,
 	std::vector<PointedThing> &objects)
 {
-	std::vector<u16> objectIds;
+	std::vector<ServerActiveObject *> objectIds;
```
Comment (sfan5) [inline src/serverenvironment.cpp:1606]: should be renamed to `objects` too

File: src/collision.cpp (hunk 1 above)
Comment (nerzhul) [reply src/collision.cpp:366]: unfortunately not because it's a parent class, compiler refuses it :(

File: src/serverenvironment.cpp (hunk 4 above)
Comment (nerzhul) [reply src/serverenvironment.cpp]: renamed to objs, we have objects as a parameter already

**Comment by SmallJoker** [comment]
Found fitting test mod (N° 02/65): https://github.com/minetest/minetest/pull/6803#issuecomment-353639173
Addition in collision.cpp L360: `ScopeProfiler sp(g_profiler, "Objects in radius", SPT_AVG);`
Addition in test mod: `
minetest.registered_entities["__builtin:item"].initial_properties.collide_with_objects = true`
Test case: `/a 3`

PR + profiler: 82us (8.2e-5)
master + profiler: 142us (0.000142)
Each ~3000 calls/second
Regarding the collision code (only), I measured a ~40% speed improvement.

File: src/serverenvironment.h [hunk 5]
```diff
@@ -323,9 +323,10 @@ class ServerEnvironment : public Environment
 	bool swapNode(v3s16 p, const MapNode &n);
 
 	// Find all active objects inside a radius around a point
-	void getObjectsInsideRadius(std::vector<u16> &objects, const v3f &pos, float radius)
+	void getObjectsInsideRadius(std::vector<ServerActiveObject *> &objects, const v3f &pos, float radius,
+			std::function<bool(ServerActiveObject *obj)> includeObjCallback)
```
Comment (sfan5) [inline src/serverenvironment.h:327]: snake_case

Comment (nerzhul) [reply src/serverenvironment.h:327]: ah i missed that one :)

**Comment by nerzhul** [comment]
> Found fitting test mod (N° 02/65): [#6803 (comment)](https://github.com/minetest/minetest/pull/6803#issuecomment-353639173)
> Addition in collision.cpp L360:
> 
>     * `ScopeProfiler sp(g_profiler, "Objects in radius", SPT_AVG);`
> 
> 
> Addition in the linked test mod:
> 
>     * ` minetest.registered_entities["__builtin:item"].initial_properties.collide_with_objects = true`
> 
> 
> Test command: `/a 3`
> 
> PR + profiler: 82us (8.2e-5)
> master + profiler: 142us (0.000142)
> Each ~3000 calls/second
> Regarding the collision code (only), I measured a ~40% speed improvement.

thanks for the tests @SmallJoker

**Pull request merged**